In [62]:
install.packages('tidyverse')
install.packages('vtable')
install.packages('naniar')


Updating HTML index of packages in '.Library'

Making 'packages.html' ...
 done

Updating HTML index of packages in '.Library'

Making 'packages.html' ...
 done

Updating HTML index of packages in '.Library'

Making 'packages.html' ...
 done



In [63]:
library(tidyverse)
library(vtable)
library(naniar)

In [64]:
marketing_campaign <- read_tsv("https://raw.githubusercontent.com/wend1k3/CWL2-STAT-301/refs/heads/main/marketing_campaign.csv")
num_observations <- nrow(marketing_campaign)
num_variables <- ncol(marketing_campaign)
name_type_table <- data.frame(
    variable = names(marketing_campaign),
    type = sapply(marketing_campaign, class))
row.names(name_type_table) <- NULL

Rows: 2240 Columns: 29
── Column specification ────────────────────────────────────────────────────────
Delimiter: "\t"
chr  (3): Education, Marital_Status, Dt_Customer
dbl (26): ID, Year_Birth, Income, Kidhome, Teenhome, Recency, MntWines, MntF...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


In [65]:
print(paste('There are ', num_observations, ' observations, and ', num_variables, ' variables in the dataset.'))
kable(name_type_table) #data types

[1] "There are  2240  observations, and  29  variables in the dataset."




|variable            |type      |
|:-------------------|:---------|
|ID                  |numeric   |
|Year_Birth          |numeric   |
|Education           |character |
|Marital_Status      |character |
|Income              |numeric   |
|Kidhome             |numeric   |
|Teenhome            |numeric   |
|Dt_Customer         |character |
|Recency             |numeric   |
|MntWines            |numeric   |
|MntFruits           |numeric   |
|MntMeatProducts     |numeric   |
|MntFishProducts     |numeric   |
|MntSweetProducts    |numeric   |
|MntGoldProds        |numeric   |
|NumDealsPurchases   |numeric   |
|NumWebPurchases     |numeric   |
|NumCatalogPurchases |numeric   |
|NumStorePurchases   |numeric   |
|NumWebVisitsMonth   |numeric   |
|AcceptedCmp3        |numeric   |
|AcceptedCmp4        |numeric   |
|AcceptedCmp5        |numeric   |
|AcceptedCmp1        |numeric   |
|AcceptedCmp2        |numeric   |
|Complain            |numeric   |
|Z_CostContact       |numeric   |
|Z_Revenue  

### Summary of Data

Note: The data types of the variables in the raw data might not be the best representation of their semantic meanings or suitable for specific models. I will change them as needed during wrangling.


|Variable            |Type      | Description|
|:-------------------|:---------|:---|
|ID                  |numeric   | Unique ID |
|Year_Birth          |numeric   | Year of birth|
|Education           |character | Education level |
|Marital_Status      |character | Marital status |
|Income              |numeric   | Yearly income|
|Kidhome             |numeric   | Number of children in households|
|Teenhome            |numeric   | Number of teenagers in household|
|Dt_Customer         |character | Date of enrollment with company|
|Recency             |numeric   | Number of days since the last purchase|
|MntWines            |numeric   | Amount spent on wine in the previous 2 years|
|MntFruits           |numeric   | Amount spent on fruits in the previous 2 years|
|MntMeatProducts     |numeric   | Amount spent on meat in the previous 2 years|
|MntFishProducts     |numeric   | Amount spent on fish in the previous 2 years|
|MntSweetProducts    |numeric   | Amount spent on sweets in the previous 2 years|
|MntGoldProds        |numeric   | Amount spent on gold in the previous 2 years|
|NumDealsPurchases   |numeric   | Number of purchases made with a discount|
|NumWebPurchases     |numeric   | Number of purchases made through the website|
|NumCatalogPurchases |numeric   | Number of purchases made through the catalogue|
|NumStorePurchases   |numeric   | Number of purchases made in store|
|NumWebVisitsMonth   |numeric   | Number of website visits in the last month|
|AcceptedCmp3        |numeric   | Whether the customer accepted a promotion offer in the 3rd campaign - 1 if yes, 0 if no|
|AcceptedCmp4        |numeric   | Whether the customer accepted a promotion offer in the 4th campaign - 1 if yes, 0 if no|
|AcceptedCmp5        |numeric   | Whether the customer accepted a promotion offer in the 5th campaign - 1 if yes, 0 if no|
|AcceptedCmp1        |numeric   | Whether the customer accepted a promotion offer in the 1st campaign - 1 if yes, 0 if no|
|AcceptedCmp2        |numeric   | Whether the customer accepted a promotion offer in the 2nd campaign - 1 if yes, 0 if no|
|Complain            |numeric   | Whether the customer complained in the last 2 years - 1 if yes, 0 otherwise|
|Z_CostContact       |numeric   | No information provided|
|Z_Revenue           |numeric   | No information provided|
|Response            |numeric   | Whether the customer accepted a promotion offer in the most recent campaign - 1 if yes, 0 if no|


### Data Collection

There is no information on Kaggle regarding how the data was collected. Given the structure of the data, I will assume that the data comes from a company collecting data on its customers. This data is likely a small sample of the entire company's records or of a specific time period, given its relatively small size.

### Data Source
The data is from kaggle, by the user **imakash3011** at this link: https://www.kaggle.com/datasets/imakash3011/customer-personality-analysis/data. The description in the data card notes that the dataset is provided by Dr. Omar Romero-Hernandez. There is no specific citation requested by the author, and the dataset is listed under a public domain license. 

An APA format citation of this dataset is:

imakash3011. (n.d.). _Customer Personality Analysis_. Kaggle. Retrieved March 14, 2026, from https://www.kaggle.com/datasets/imakash3011/customer-personality-analysis

To read in the data, I have used a link from my group's GitHub repository.

## Exploratory Data Analysis and Visualization

In [66]:
#reading and loading data
marketing_campaign <- read_tsv("https://raw.githubusercontent.com/wend1k3/CWL2-STAT-301/refs/heads/main/marketing_campaign.csv")
head(marketing_campaign, 10)

Rows: 2240 Columns: 29
── Column specification ────────────────────────────────────────────────────────
Delimiter: "\t"
chr  (3): Education, Marital_Status, Dt_Customer
dbl (26): ID, Year_Birth, Income, Kidhome, Teenhome, Recency, MntWines, MntF...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


ID,Year_Birth,Education,Marital_Status,Income,Kidhome,Teenhome,Dt_Customer,Recency,MntWines,⋯,NumWebVisitsMonth,AcceptedCmp3,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Z_CostContact,Z_Revenue,Response
<dbl>,<dbl>,<chr>,<chr>,<dbl>,<dbl>,<dbl>,<chr>,<dbl>,<dbl>,⋯,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>
5524,1957,Graduation,Single,58138,0,0,04-09-2012,58,635,⋯,7,0,0,0,0,0,0,3,11,1
2174,1954,Graduation,Single,46344,1,1,08-03-2014,38,11,⋯,5,0,0,0,0,0,0,3,11,0
4141,1965,Graduation,Together,71613,0,0,21-08-2013,26,426,⋯,4,0,0,0,0,0,0,3,11,0
6182,1984,Graduation,Together,26646,1,0,10-02-2014,26,11,⋯,6,0,0,0,0,0,0,3,11,0
5324,1981,PhD,Married,58293,1,0,19-01-2014,94,173,⋯,5,0,0,0,0,0,0,3,11,0
7446,1967,Master,Together,62513,0,1,09-09-2013,16,520,⋯,6,0,0,0,0,0,0,3,11,0
965,1971,Graduation,Divorced,55635,0,1,13-11-2012,34,235,⋯,6,0,0,0,0,0,0,3,11,0
6177,1985,PhD,Married,33454,1,0,08-05-2013,32,76,⋯,8,0,0,0,0,0,0,3,11,0
4855,1974,PhD,Together,30351,1,0,06-06-2013,19,14,⋯,9,0,0,0,0,0,0,3,11,1


#### Cleaning and Wrangling
The data is already in a tidy format, but I need to change the formatting of some of the data types and parse the dates, as well as drop the irrelevant columns.

#### Unused Variables
The variables `Z_CostContact` and `Z_Revenue` have a constant value of 3 and 11, respectively. Additionally, there is no documention describing the information they represet. Thus, we have chosen to omit them in our analysis.

The variable `ID` is used to represent unique customer IDs. This is not relevant to our analysis because it does not provide any relevant information. So we will omit it as well.

In [67]:
#dropping irrelevant variables
marketing_campaign <- marketing_campaign %>%
    select(-ID, -Z_CostContact, -Z_Revenue)

#### Missing Values

In [68]:
miss_var_summary(marketing_campaign)

variable,n_miss,pct_miss
<chr>,<int>,<num>
Income,24,1.07
Year_Birth,0,0
Education,0,0
Marital_Status,0,0
Kidhome,0,0
Teenhome,0,0
Dt_Customer,0,0
Recency,0,0
MntWines,0,0


There are 24 missing values under `Income`, with no other missing values under the other variables. Since there are a total of 2240 observations, which only makes up of 1.07% of the data. Thus, we simply remove the 24 rows with those missing values.

In [69]:
marketing_campaign_dropna <- marketing_campaign %>% drop_na(Income)
num_observations_dropna <- nrow(marketing_campaign_dropna)
num_variables_dropna <- ncol(marketing_campaign_dropna)
print(paste('After dropping unused columns and observations with missing valees, there are now ', num_observations_dropna, ' observations, and ', num_variables_dropna, ' variables in the dataset.'))

[1] "After dropping unused columns and observations with missing valees, there are now  2216  observations, and  26  variables in the dataset."


#### Data Types
We will make sure that all the data types accurately reflect their semantic meanings. This is important for interpretation and also to ensure they are appropriate to use as model inputs. 

##### Factor Levels: `Education`
Education is generally considered an ordered variable. We will inspect the unique levels to make sure they are in order.


In [70]:
unique(marketing_campaign$Education) #checking to create the correct ordering

[1] "Graduation" "PhD"        "Master"     "Basic"      "2n Cycle"

`2n Cycle` is likely referring to european second cycle programs, which are post-graduate degrees. We will put them in between `Graduation` and `Master`.

In [71]:
marketing_campaign <- marketing_campaign %>% 
    mutate(
        Education = factor(Education, 
                            levels = c('Basic','Graduation', '2n Cycle', 'Master', 'PhD'),
                            ordered = TRUE) #making education a ordered factor
    )

##### Factor Levels: `Marital_Status`
`Marital_Status` should be a factor. Before confirming its data type, we will inspect the unique levels to make sure there are no extreme outliers or mistakes. 

In [72]:
data.frame(table(marketing_campaign_dropna$Marital_Status)) %>% 
    rename(Count = Freq) %>% 
    mutate(Proportion = prop.table(Count))

Var1,Count,Proportion
<fct>,<int>,<dbl>
Absurd,2,0.0009025271
Alone,3,0.0013537906
Divorced,232,0.1046931408
Married,857,0.3867328520
Single,471,0.2125451264
Together,573,0.2585740072
Widow,76,0.0342960289
YOLO,2,0.0009025271


It seems like `Absurd`, `Alone`, and `YOLO` are outliers, with only 2-3 observations each. This could be problematic, because the `Marital_Status` variable is categorical and also probably relevant to the question. I think it's safe to assume that `Alone` and `YOLO` are similar enough to `Single`, so I will recode them. `Absurd` does not quite fit in with any of the common categories, and it seems like it might have been entered as a joke or a typo. Given how few observations there are and the fact that it does not fit in with anything else, I will drop it.

In [73]:
marketing_campaign<- marketing_campaign %>% 
    mutate(
         Marital_Status = as.factor(Marital_Status)) %>% #making marital status a factor
    filter(Marital_Status != 'Absurd') %>% #dropping extremely rare level
    droplevels() %>% #dropping unused level
    mutate(Marital_Status = fct_collapse(Marital_Status, Single = c('Single', 'Alone', 'YOLO'))) #combining "Alone" and "YOLO" with single, at our discretion 
data.frame(table(marketing_campaign_clean$Marital_Status)) %>% #displaying new results after wrangling
    rename(Count = Freq) %>% 
    mutate(Proportion = prop.table(Count))

Var1,Count,Proportion
<fct>,<int>,<dbl>
Single,475,0.21493213
Divorced,231,0.10452489
Married,857,0.38778281
Together,571,0.25837104
Widow,76,0.03438914


Now, we can convert the data types in our data to suit our needs. 

What we have already done: 
- convert `Education` to ordered factor
- convert `Marital_Status` to factor

What we still need to do: 
- make sure that the date variables (`Year_Birth` and `Dt_Customer`) are properly parsed
- make sure the binary response variables are cast as factors
- make sure all the numeric variables (recency, income, number of kids/teens, amounts spent on products; number of purchases/visits) are cast as numeric variables. 

In [74]:
# data cleaning and wrangling into a tidy format
marketing_campaign <- marketing_campaign %>%
    mutate(
           Year_Birth = as.numeric(Year_Birth),
           Income = as.numeric(Income),
           Kidhome = as.numeric(Kidhome),
           Teenhome = as.numeric(Teenhome),
           Dt_Customer = as.Date(Dt_Customer, format="%d-%m-%Y"),
           Recency = as.numeric(Recency),
           Complain = as.factor(Complain),
           MntWines = as.numeric(MntWines),
           MntFruits = as.numeric(MntFruits),
           MntMeatProducts = as.numeric(MntMeatProducts),
           MntFishProducts = as.numeric(MntFishProducts),
           MntSweetProducts = as.numeric(MntSweetProducts),
           MntGoldProds = as.numeric(MntGoldProds),
           NumDealsPurchases = as.numeric(NumDealsPurchases),
           AcceptedCmp1 = as.numeric(AcceptedCmp1),
           AcceptedCmp2 = as.numeric(AcceptedCmp2),
           AcceptedCmp3 = as.numeric(AcceptedCmp3),
           AcceptedCmp4 = as.numeric(AcceptedCmp4),
           AcceptedCmp5 = as.numeric(AcceptedCmp5),
           Response = as.factor(Response),
           NumWebPurchases = as.numeric(NumWebPurchases),
           NumCatalogPurchases = as.numeric(NumCatalogPurchases),
           NumStorePurchases = as.numeric(NumStorePurchases),
           NumWebVisitsMonth = as.numeric(NumWebVisitsMonth))
head(marketing_campaign)

Year_Birth,Education,Marital_Status,Income,Kidhome,Teenhome,Dt_Customer,Recency,MntWines,MntFruits,⋯,NumCatalogPurchases,NumStorePurchases,NumWebVisitsMonth,AcceptedCmp3,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Response
<dbl>,<ord>,<fct>,<dbl>,<dbl>,<dbl>,<date>,<dbl>,<dbl>,<dbl>,⋯,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<fct>,<fct>
1957,Graduation,Single,58138,0,0,2012-09-04,58,635,88,⋯,10,4,7,0,0,0,0,0,0,1
1954,Graduation,Single,46344,1,1,2014-03-08,38,11,1,⋯,1,2,5,0,0,0,0,0,0,0
1965,Graduation,Together,71613,0,0,2013-08-21,26,426,49,⋯,2,10,4,0,0,0,0,0,0,0
1984,Graduation,Together,26646,1,0,2014-02-10,26,11,4,⋯,0,4,6,0,0,0,0,0,0,0
1981,PhD,Married,58293,1,0,2014-01-19,94,173,43,⋯,3,6,5,0,0,0,0,0,0,0
1967,Master,Together,62513,0,1,2013-09-09,16,520,42,⋯,4,10,6,0,0,0,0,0,0,0


#### Class Imbalance
Now we will examine any class imbalance in our data. We have two categorical variables to consider: `Education` and `Marital_Status`. And we have seven binary variables to consider: `AcceptedCmp` 1 through 5, `Complain`, and `Respose`.

First, let's look at `Marital_Status`.

In [75]:
data.frame(table(marketing_campaign$Marital_Status)) %>% 
    rename(Count = Freq) %>% 
    mutate(Proportion = prop.table(Count))

Var1,Count,Proportion
<fct>,<int>,<dbl>
Single,485,0.21671135
Divorced,232,0.10366399
Married,864,0.38605898
Together,580,0.25915996
Widow,77,0.03440572


These results are as expected. `Widow` is quite rare, with only about 3% of the observations. However, this makes sense, as widows are relatively more rare than the other levels in practice. This is not a major concern unless the `Widow` level ends up being very significant in our model.

I will also take a look at `Education`, the other categorical variable.

In [76]:
data.frame(table(marketing_campaign$Education)) %>% 
    rename(Count = Freq) %>% 
    mutate(Proportion = prop.table(Count))

Var1,Count,Proportion
<fct>,<int>,<dbl>
Basic,54,0.02412869
Graduation,1126,0.50312779
2n Cycle,203,0.09070599
Master,369,0.16487936
PhD,486,0.21715818


Education looks alright. `Basic` is quite rare, but there are a decent amount of observations, and it makes sense given the context; most people have more than a basic level of education. Again, this is not too concerning.

Next, I will take a look at all the binary response variables. 

In [77]:
binary_responses <- marketing_campaign_clean %>% 
    select(AcceptedCmp3, AcceptedCmp4, AcceptedCmp5, 
           AcceptedCmp1, AcceptedCmp2, Complain, Response) %>%
    pivot_longer(cols = everything()) %>% 
    group_by(name) %>%
    summarize(count_1s = sum(value == 1),
              count_0s = sum (value == 0),
              percent_1s = (100 * mean(value)))
    

binary_responses

name,count_1s,count_0s,percent_1s
<chr>,<int>,<int>,<dbl>
AcceptedCmp1,141,2069,6.3800905
AcceptedCmp2,30,2180,1.3574661
AcceptedCmp3,163,2047,7.3755656
AcceptedCmp4,164,2046,7.4208145
AcceptedCmp5,160,2050,7.2398190
Complain,20,2190,0.9049774
Response,332,1878,15.0226244


There is some class imbalance in that 1's (positive responses) are much less common than 0's, but the class imbalance is not too severe in most cases, and it's likely an accurate reflection of real-life behaviour. However, `Complain` and `AcceptedCmp2` have concerningly few positive observations. We will be cautious about using these in a model and interpreting their results. 

A possible solution for this is to create some engineered features. We will address this in a later section.

#### Outliers
Given that some variables represent real-life demographic information, I will check for 'unreasonable' values.

In [78]:
sumtable(marketing_campaign_clean, out = 'return')

Variable,N,Mean,Std. Dev.,Min,Pctl. 25,Pctl. 75,Max
<chr>,<chr>,<chr>,<chr>,<chr>,<chr>,<chr>,<chr>
Year_Birth,2210,1969,12,1940,1959,1977,1996
Education,2210,,,,,,
... Basic,54,2%,,,,,
... 2n Cycle,198,9%,,,,,
... Graduation,1114,50%,,,,,
... Master,364,16%,,,,,
... PhD,480,22%,,,,,
Marital_Status,2210,,,,,,
... Single,475,21%,,,,,


I am concered with `Year_Birth` and `Income`. There seems to be a very early year and a very high income. 

In [79]:
earliest_year_birth <- head(sort(marketing_campaign_clean$Year_Birth, decreasing = FALSE), n = 5)
highest_income <- head(sort(marketing_campaign_clean$Income, decreasing = TRUE), n = 5)

print(paste('Earliest Birth Years: ', list(earliest_year_birth)))
print(paste('Highest Incomes: ', list(highest_income)))


[1] "Earliest Birth Years:  c(1940, 1941, 1943, 1943, 1943)"
[1] "Highest Incomes:  c(162397, 160803, 157733, 157243, 157146)"


The list of earliest birth years seems a little problematic. 1893, 1899, and 1900 are very early birth years. Perhaps they are due to misinputs or customers entering random birth years. Because there are only 3 concerning values, I will just drop them by setting the cutoff at 100 years ago. The very highest income also seems problematic--it is much higher than the next value down, and it's the same digit repeated. Perhaps it's a typo. Again, seeing as there is only one extreme outlier and it is likely an input error, I will just drop it by cutting income off at 200,000. 

In [80]:
marketing_campaign_clean <- marketing_campaign_clean %>% filter(Year_Birth >= (year(now()) - 100 ),
                                                                Income <= 200000)

#### Feature Engineering

We have decided to engineer three features: 

1. `Year_Birth` -> `Age`
This aligns with inutitive interpretaitons regarding age (higher = older, rather than higher = younger). 
Note: the dataset was uplaoded in 2021, and has not been updated since. Thus, we will use 2021 to calculate ages.

2. `Dt_Customer` -> `Recency`
We will use the most recent customer enrollment date to compute the amount of time a customer has been enrolled. This avoids issues with models treating the datetime variable as continuous.

3. `AcceptedCmp1-5` -> `AcceptedCmpAny`
To make sure our analysis is not significantly affected by class imbalance, we will combine all the campaign acceptance variables, with 1 representing that the customer has accepted any of the 5 campaigns.

We will drop the original variables with which we engineered the new variables to avoid multicollinearity. 

In [81]:
marketing_campaign <- marketing_campaign %>% 
    mutate(
        Age = 2021 - Year_Birth,
        Recency = as.numeric(max(marketing_campaign$Dt_Customer) - Dt_Customer),
        AcceptedCmpAny = if_else(
            (AcceptedCmp1 == 1 | 
            AcceptedCmp2 == 1 | 
            AcceptedCmp3 == 1 |
            AcceptedCmp4 == 1 |
            AcceptedCmp5 == 1),
            1,
            0)
        ) %>%
    select(
        -Year_Birth, 
        -Dt_Customer, 
        -AcceptedCmp1, 
        -AcceptedCmp2, 
        -AcceptedCmp3, 
        -AcceptedCmp4,
        -AcceptedCmp5)

    

### Visualizations
